### PR TITLE
Add RFC 102 text: Embedding resource files into libgdal

### PR DESCRIPTION
**Pull request by rouault**
[Rendered view](https://gdal--10913.org.readthedocs.build/en/10913/development/rfc/rfc102_embedded_resources.html)

### CODE REVIEW

**Comment by sgillies** [comment]
@rouault I'm using CPLFindFile() to detect whether a GDAL install is inside a Python package: https://github.com/rasterio/rasterio/blob/df409de086efa437fef2b66b4e0d2e1e387fe474/rasterio/_env.pyx#L250. Could that function be updated to search in libgdal? Or should I depend on not using this option?

**Comment by rouault** [comment]
> Could that function be updated to search in libgdal?

no, embedded resources are not easily discoverables as sub-files. That would involve complicated low-level compiler/OS-specific trick

> Or should I depend on not using this option?

Well, the intent is that your GDAL builds for Rasterio/Fiona binary wheels do use the new CMake option EMBED_RESSOURCE_FILES=ON, and probably also USE_ONLY_EMBEDDED_RESOURCE_FILES=ON, since it is unlikely users would want to tune the default resource files.

**Comment by rouault** [comment]
FYI @Maxxen w.r.t proj.db embedding

**Comment by eschnett** [comment]
This feature could easily be implemented in Perl or Python with some makefile support, generating a C source file from a binary that can then be `#include`d whenever `#embed` is not available.

We have been using this feature in the [Einstein Toolkit](https://einsteintoolkit.org) for many years (hence Perl) in the [Formaline](https://bitbucket.org/cactuscode/cactusutils/src/master/Formaline/) thorn. This code does many other things as well; the actual code is in the `makeblob.pl` and `makemetablob.pl` Perl files in the `src/util` directory.

One important thing at the time (decades ago) was to ensure that the generated source files are not too large since certain compilers would run out of memory with large literal arrays. That may or may not be necessary any more.

**Comment by rouault** [comment]
> This feature could easily be implemented in Perl or Python with some makefile support

yes, that's what I mention in the "alternatives" paragraph. But I don't want to implement both C23 #embed support, and older tricks. Too painful too maintain. Let's go to the future straight ahead.

> One important thing at the time (decades ago) was to ensure that the generated source files are not too large since certain compilers would run out of memory with large literal arrays. That may or may not be necessary any more.

GCC 15dev #embed unoptimized implementation shows that's still a concern. CLang 19 #embed implementation uses collaboration between the pre-processor and later stages of the compilation to have time and RAM efficient embedding.

**Comment by jjimenezshaw** [comment]
Is there going to be a similar option only in PROJ? (Without complling GDAL)

**Comment by rouault** [comment]
> Is there going to be a similar option only in PROJ? (Without complling GDAL)

yes

**Comment by Maxxen** [comment]
One thing that's on my todo list is investigating using sqlites memvfs in combination with its [zipvfs](https://sqlite.org/zipvfs/doc/trunk/www/howitworks.wiki) file system, potentially reducing the size of the embedded PROJ database even further.

File: doc/source/development/rfc/rfc102_embedded_resources.rst [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+an issue for GDAL use cases, and there is intent from GCC developers to improve
+this in the future.
+
+Embedding PROJ's :file:`proj.db` of size 9.1 MB with GCC 15dev at time of writing takes
+18 seconds and 1.7 GB RAM, compared to 0.4 second and 400 MB RAM for clang 19,
+which is still reasonable (Generating :file:`proj.db` itself from its source .sql files
+takes one minute on the same system).
+
+There is no timeline for Visual Studio C/C++ at time of writing (it has been
+`requested by users <https://developercommunity.visualstudio.com/t/Add-support-for-embed-as-voted-into-the/10451640?ru=https%3A%2F%2Fdevelopercommunity.visualstudio.com%2Fcontent%2Fproblem%2F132967%2Fcopycut-commands-doesnt-work.html%3FchildToView%3D133119>`__)
```
Comment (schwehr) [inline doc/source/development/rfc/rfc102_embedded_resources.rst:69]: Can the URL be shortened to https://developercommunity.visualstudio.com/t/Add-support-for-embed-as-voted-into-the/10451640 ?

Comment (rouault) [reply doc/source/development/rfc/rfc102_embedded_resources.rst]: done, thanks

**Comment by rouault** [comment]
PROJ specificities are now covered by a dedicated PROJ RFC-8: https://github.com/OSGeo/PROJ/pull/4274

File: doc/source/development/rfc/rfc102_embedded_resources.rst [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+to be able to optionally embed GDAL resource files directly into libgdal.
+
+A similar `PROJ RFC-8 <https://github.com/OSGeo/PROJ/pull/4274>`__ has been
+submitted for PROJ to embed its :file:`proj.db` and :file:`proj.ini` files.
+
+Motivation
+----------
+
+Some parts of GDAL core, but mostly drivers, depend on a number of resource
+files for correct execution. Locating those resource files on the filesystem
```
Comment (pjonsson) [inline doc/source/development/rfc/rfc102_embedded_resources.rst:28]: I find parts of this paragraph difficult to read. Does this convey the meaning of the first two sentences:

".. mostly drivers, require external resource files located in the filesystem. Locating these resource files is difficult for use cases where the GDAL binaries are relocated during installation time."

Take this with a grain of salt though, English isn't my first language.

File: doc/source/development/rfc/rfc102_embedded_resources.rst [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+can be painful in some use cases of GDAL, that involve relocating the GDAL
+binary at installation time. One such case could be the GDAL embedded in Rasterio
+or Fiona binary wheels where :config:`GDAL_DATA` must be correctly set currently.
```
Comment (pjonsson) [inline doc/source/development/rfc/rfc102_embedded_resources.rst:31]: Does replacing "must be correctly set currently" with "must be set to the directory of the resource files" work?

File: doc/source/development/rfc/rfc102_embedded_resources.rst (hunk 2 above)
Comment (rouault) [reply doc/source/development/rfc/rfc102_embedded_resources.rst]: thanks, adopted

File: doc/source/development/rfc/rfc102_embedded_resources.rst (hunk 3 above)
Comment (rouault) [reply doc/source/development/rfc/rfc102_embedded_resources.rst]: thanks, adopted

File: doc/source/development/rfc/rfc102_embedded_resources.rst [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+===================================================================
+RFC 102: Embedding resource files into libgdal
+===================================================================
+
+============== =============================================
+Author:        Even Rouault
+Contact:       even.rouault @ spatialys.com
+Started:       2024-Oct-01
+Status:        Draft
```
Comment (rouault) [inline doc/source/development/rfc/rfc102_embedded_resources.rst:11]:
```suggestion
Status:        Adopted, implemented
```

File: doc/source/development/rfc/rfc102_embedded_resources.rst [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+- https://github.com/OSGeo/gdal/issues/10780
+
+- `GDAL candidate implementation <https://github.com/OSGeo/gdal/pull/10972>`__
+
+- `PROJ RFC-8 Embedding resource files into libproj <https://github.com/OSGeo/PROJ/pull/4274>`__
+
+Voting history
+--------------
+
+TBD
```
Comment (rouault) [inline doc/source/development/rfc/rfc102_embedded_resources.rst:165]:
```suggestion
+1 from PSC members JukkaR, JavierJS, KurtS, HowardB and EvenR
```

**Pull request merged**